JS: Static initializer blocks and "for await ... of ..."

matthiasblaesing opened the pull request:
This PR adds support for parsing static initializer blocks:

```javascript
class Example{
  static attrib;
  static {
   this.attrib = "exampleValue";
  }
}
```

and "for await ... of ..." loops:

```javascript
async function* foo() {
  yield 1;
  yield 2;
}

(async function() {
  for await (const num of foo()) {
    console.log(num);   // expected output: 1
    break;              // closes iterator, triggers return
  }
})();
```

Closes: #4262
Closes: #4757

matthiasblaesing commented:
The first two commits are identical to #5677 and that PR should be merged first.

Special attention is advised for the commits f17e17d, 0b818e0 and e7a40ce. These commits import/regenerate code, that was externalized in the donation process. While Oracle was indeed not in the situation to donate the code, Apache projects can import code under MIT/BSD and UPL licenses.

The JS parser is based on GraalJS, which was later relicensed to UPL. @jtulach rebased the NetBeans specific changes onto that branch, which resulted in a uniformly UPL licensed codebase. That codebase was then further updated by me to cover more JS features. The resulting binary was already part of the release, the change is here, that now the code is also in the project.

The testfiles for the javascript2.editor were taken from external projects and licensed under the BSD/MIT license. Both can be incorporated into Apache projects, which is done now. This was necessary as the AST structure changed and thus the goldenfiles had to be regenerated.

vieiro commented on webcommon/libs.nashorn/licenseinfo.xml:
```diff
@@ -0,0 +1,129 @@
+        <file>test/unit/src/com/oracle/js/parser/DumpingVisitor.java</file>
+        <file>test/unit/src/com/oracle/js/parser/ManualTest.java</file>
+        <file>test/unit/src/com/oracle/js/parser/ParserTest.java</file>
+        <license ref="UPL" />
```
Do we want to use [UPL-Apache-2.0](https://github.com/apache/netbeans/blob/master/nbbuild/licenses/UPL-Apache-2.0) instead? 

Or we may even want to create a new UPL-Oracle license:

- The "UPL" license has a generic copyright information.
- UPL-Apache-2.0 includes Oracle's copyright, but for years 2016-2020.
- We are importing files with Oracle copyright for 2010-2018, though.

matthiasblaesing commented:
@vieiro thank you for checking. I pushed an update with three more changes (to be partially squashed with their base commits):

- I created a new formatted version of the graaljs license with the year range covering the files that are part of the first integration commit
- I added a second commit, that also covers the new `ClassElement.java` file imported, to be squashed with the corresponding commit
- I relicensed the tests to ALv2 and removed the UPL header as the author I'm in the position to do that.

This should improve the situation.

matthiasblaesing commented:
@jtulach thank for looking into this and no reason to apologize. The current code that is proposed to be integrated here has already an outside history, ontop of your work. I looked into the divergence between the upstream graaljs and the version you thankfully created and faced to much for me to handle. So I created my own fork (https://github.com/matthiasblaesing/graaljs-nb) and modified the parser in a manor I could manage to support new JS features. This was also motivated by the fact, that graaljs upstream does not support JSX and thus a move fully to upstream would kill that feature in NetBeans.

We can try to get the fork more in-line with upstream again and reduce differences, but I don't see us switching back to upstream again. So I agree with you and @vieiro, that better alignment would be good, I also have to be realistic what I can archive.

matthiasblaesing commented:
@jtulach for the "Can it be integrated?" question: The ASF 3rd party license policy lists the `UPL` as Category-A ("For inclusion in an Apache Software Foundation product, we consider the following licenses to be similar in terms to the Apache License 2.0") https://www.apache.org/legal/resolved.html#category-a).

I'd like to integrate as is, do you object or is it ok from your perspective?

matthiasblaesing commented:
I did the squashing and will merge tomorrow once the tests are green.

paulongithub commented:
I still have this error in Netbeans 17. Is there something I need to do to the IDE to make it recognise `for await` in an `async` function? Note, code runs fine. It's just showing the row and file in error in the IDE.

matthiasblaesing commented:
This is tagged NB18, it is not part of NB17.
